File: Testes/JogoImp.java
import java.util.Random;
import static java.lang.Thread.*;

interface Jogo {
    Partida participa() throws InterruptedException;
}

interface Partida {
    String adivinha(int n) throws InterruptedException;
}

class JogoImpl implements Jogo {
    private PartidaImpl partida = new PartidaImpl();
    private int jogadores = 0;

    public synchronized Partida participa() throws InterruptedException {
        jogadores++;
        PartidaImpl p = partida;

        if(jogadores == 4) {
            notifyAll();
            jogadores = 0;
            partida = new PartidaImpl();     //nova partida para os proximos

            new Thread(() -> {              //conta os 60 segundos ate disparar
                try {
                    sleep(6000);
                } catch (InterruptedException e) {
                    p.timeout();
                }
            }).start();
        }
        else{
            while(p == partida) wait();
        }
        return p;
    }
}

class PartidaImpl implements Partida {
    int tentativas = 100;
    int numero = new Random().nextInt(100);
    boolean ganhou = false;
    boolean timeout = false;

    synchronized void timeout() {
        timeout = true;
    }

    public synchronized String adivinha(int n) {
        if (ganhou) return "PERDEU";
        else if (tentativas >= 100) return "TENTATIVAS";
        else if (timeout) return "TEMPO";
        else {
            tentativas++;
            if (n > numero) return "MENOR";
            else if (n < numero) return "MAIOR";
            else {
                ganhou = true;
                return "GANHOU";
            }
        }
    }
}

File: Testes/jogoImp.erl
-module(jogoImp).
-export([start/0, participa/1, adivinha/2]).

% Inicia o servidor do jogo
start() -> spawn(fun() -> jogo([]) end).

% Jogador participa de um jogo
participa(Jogo) -> 
    Jogo ! {participa, self()},     % Enviamos o seu pid para garantir que a resposta vem mesmo para nós
    receive 
        {Partida, Jogo} -> Partida 
    end.

% Jogador faz uma tentativa de adivinhação
adivinha(Partida, Numero) ->
    Partida ! {adivinha, Numero, self()},
    receive 
        {Res, Partida} -> Res 
    end.

% Gerencia o estado do jogo aguardando jogadores
jogo(Jogadores) when length(Jogadores) < 4 ->
    receive 
        {participa, From} ->
            jogo([From | Jogadores])
    end;

% Quando há jogadores suficientes, inicia uma nova partida
jogo(Jogadores) ->              
    Numero = random:uniform(100),
    Partida = spawn(fun() -> partida(Numero, 1, false, false) end),
    [Jogador ! {Partida, self()} || Jogador <- Jogadores],
    spawn(fun() -> receive after 60000 -> Partida ! timeout end end),    % Após um minuto manda mensagem à partida para ela terminar
    jogo([]).

% Gerencia o estado de uma partida
partida(Numero, Tentativas, Timeout, Ganhou) ->
    receive
        {adivinha, N, From} -> 
            Res =
            if
                Ganhou -> "PERDEU";
                Timeout -> "TEMPO";
                Tentativas >= 100 -> "TENTATIVAS";
                Numero < N -> "MENOR";
                Numero > N -> "MAIOR";
                true -> "GANHOU"
            end,
            From ! {Res, self()},
            partida(Numero, Tentativas + 1, Timeout, Ganhou orelse Res =:= "GANHOU");
        timeout -> 
            partida(Numero, Tentativas, true, Ganhou)
    end.